File: MORENT.Application/Services/CarService.cs
using MORENT.Application.Common;
using MORENT.Application.DTOs;
using MORENT.Application.Interfaces.Persistence;
using MORENT.Application.Interfaces.Services;
using MORENT.Domain.Entities.Dbo;

namespace MORENT.Application.Services
{
    public class CarService : ICarService
    {
        private readonly IUnitOfWork _uow;

        public CarService(IUnitOfWork uow)
        {
            _uow = uow;
        }

        public async Task<Result<IReadOnlyList<CarDto>>> GetPopularCarsAsync(int count = 4)
        {
            var cars = await _uow.Cars.GetPopularCarsAsync(count);
            return Result<IReadOnlyList<CarDto>>.Success(cars);
        }

        public async Task<Result<IReadOnlyList<CarDto>>> GetRecommendedCarsAsync(int count = 4)
        {
            var cars = await _uow.Cars.GetRecommendedCarsAsync(count);
            return Result<IReadOnlyList<CarDto>>.Success(cars);
        }

        public async Task<Result<IReadOnlyList<LocationDto>>> GetAvailableLocationsAsync()
        {
            var locations = await _uow.Cars.GetAvailableLocationsAsync();
            return Result<IReadOnlyList<LocationDto>>.Success(locations);
        }

        public async Task<Result<PagedResult<CarDto>>> GetFilteredCarsAsync(
            string? searchTerm, string? carType, int? pickUpLocationId,
            int? capacity, string? steeringType, decimal? maxPrice,
            DateTime? pickUpDate, DateTime? dropOffDate,
            int pageNumber = 1, int pageSize = 9)
        {
            if (pageNumber < 1) pageNumber = 1;
            if (pageSize < 1) pageSize = 9;
            if (pageSize > 45) pageSize = 45;

            var cars = await _uow.Cars.GetFilteredCarsAsync(searchTerm, carType, pickUpLocationId,
                capacity, steeringType, maxPrice, pickUpDate, dropOffDate, pageNumber, pageSize);
            return Result<PagedResult<CarDto>>.Success(cars);
        }

        public async Task<Result<CarDto>> GetCarDetailsAsync(Guid carId)
        {
            var car = await _uow.Cars.GetCarDetailsAsync(carId);
            if (car == null)
            {
                return Result<CarDto>.Failure("Car not found.");
            }

            return Result<CarDto>.Success(car);
        }

        public async Task<Result<IReadOnlyList<CarDto>>> GetUserFavoriteCarsAsync(Guid userId)
        {
            var cars = await _uow.Cars.GetUserFavoriteCarsAsync(userId);
            return Result<IReadOnlyList<CarDto>>.Success(cars);
        }

        public async Task<Result<bool>> ToggleFavoriteAsync(Guid userId, Guid carId)
        {
            var existingFavorite = await _uow.Cars.GetFavoriteAsync(userId, carId);

            if (existingFavorite != null)
            {
                // Unfavorite
                _uow.Cars.RemoveFavorite(existingFavorite);
                await _uow.SaveChangesAsync();

                return Result<bool>.Success(false, "Removed from favorites");
            }

            // Favorite
            var favorite = new FavoriteCar
            {
                UserId = userId,
                CarId = carId
            };

            await _uow.Cars.AddFavoriteAsync(favorite);
            await _uow.SaveChangesAsync();

            return Result<bool>.Success(true, "Added to favorites");
        }
    }
}

File: MORENT.Presentation/MORENT.Presentation.Client/src/app/features/cars/services/car.service.ts
import { Injectable, inject } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { environment } from "../../../../environments/environment";
import { IPagedResult, IResult } from "../../../core/interfaces/result";
import { ICarDto } from "../interfaces/car";
import { ILocationDto } from "../interfaces/location";

@Injectable({
  providedIn: "root",
})
export class CarService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/cars`;

  getAvailableLocations(): Observable<IResult<ILocationDto[]>> {
    return this.http.get<IResult<ILocationDto[]>>(`${this.apiUrl}/locations`);
  }

  getPopularCars(count: number = 4): Observable<IResult<ICarDto[]>> {
    return this.http.get<IResult<ICarDto[]>>(
      `${this.apiUrl}/popular?count=${count}`,
    );
  }

  getRecommendedCars(count: number = 4): Observable<IResult<ICarDto[]>> {
    return this.http.get<IResult<ICarDto[]>>(
      `${this.apiUrl}/recommended?count=${count}`,
    );
  }

  getFilteredCars(
    pageNumber: number = 1,
    pageSize: number = 9,
    pickUpLocationId?: number,
    searchTerm?: string,
    carType?: string,
    capacity?: number,
    steeringType?: string,
    maxPrice?: number,
    pickUpDate?: string,
    dropOffDate?: string,
  ): Observable<IResult<IPagedResult<ICarDto>>> {
    let params = new HttpParams()
      .set("pageNumber", pageNumber)
      .set("pageSize", pageSize);

    if (pickUpLocationId)
      params = params.set("pickUpLocationId", pickUpLocationId);
    if (searchTerm) params = params.set("searchTerm", searchTerm);
    if (carType) params = params.set("carType", carType);
    if (capacity) params = params.set("capacity", capacity);
    if (steeringType) params = params.set("steeringType", steeringType);
    if (maxPrice) params = params.set("maxPrice", maxPrice);
    if (pickUpDate) params = params.set("pickUpDate", pickUpDate);
    if (dropOffDate) params = params.set("dropOffDate", dropOffDate);

    return this.http.get<IResult<IPagedResult<ICarDto>>>(this.apiUrl, {
      params,
    });
  }

  getCarDetails(id: string): Observable<IResult<ICarDto>> {
    return this.http.get<IResult<ICarDto>>(`${this.apiUrl}/${id}`);
  }

  getFavorites(): Observable<IResult<ICarDto[]>> {
    return this.http.get<IResult<ICarDto[]>>(`${this.apiUrl}/favorites`, {
      withCredentials: true,
    });
  }

  toggleFavorite(carId: string): Observable<IResult<boolean>> {
    return this.http.post<IResult<boolean>>(
      `${this.apiUrl}/${carId}/favorite`,
      {},
      { withCredentials: true },
    );
  }
}
